### ISSUE TITLE
k2 configuration file comparison tool - spec

### ISSUE THREAD

**Issue by coffeepac** [samsung-cnct/k2#221]
changing the version of kubernetes may also mean a change in the k2 configuration file.  there should be a standalone application (written in go) or an ansible role that checks the configuration file during a kubernetes version change operation and produces the following output:
- a list of new configuration options
-- when going up in version this would be genuinely new flags
-- when going down in version this would be flags that were removed
- deprecation warnings for any options
- missing required options
- cause the version change operation to pause if any required options are missing

this spec should contain the format for storing when/how these options are added and removed.  this design should strive to be as automatic as possible:  ie try to not require that a feature developer also always update the comparison tool when a new feature is added or if the developer must take manual action there should be an automatic prompt (CI job, compile task, etc).

**Comment by coffeepac** [comment]
#222 deals with kubernetes template versioning.  if that ticket follows my currently preferred path, we can, possibly, save a copy of the k2 template in each version directory and use that for diffs.  maybe, I'm not sure if there would be required configuration file changes outside of a k8s version change but there might be.  it requires thought.

**Comment by coffeepac** [comment]
this should include removal of defaults from ansible code.

**Comment by coffeepac** [comment]
this should result in a design that says, going forward, how we are going to handle the k2 configuration file.

**Comment by joejulian** [comment]
Like @davidewatson said, try to keep config definitions as generic as possible. When defining an object, it should have the features of other objects in the parent object and be defined the same way.

For example, `master` is just `node` used for a particular purpose. Like `node`, it should contain a list.

**Comment by davidewatson** [comment]
#k2 configuration

We propose using JSON schema to validate the k2 configuration file. JSON schema provides us with a framework to partially or wholly solve a number of issues. Some of these are versioning, validation (syntax, format, defaults) and documentation. Once the configuration file is versioned our support policy will make it more difficult to change. Therefore we might take this opportunity to refactor the configuration file to take advantage of the knowledge we have gained while implementing extensions to k2.

The particulars of the configuration file will be described and iterated on later. For now we just want to agree on the framework and avoid bike shedding.

## Format

- An example (partial) configuration file:

```
---
version: v1
deployment:
  clusters:
    -
      name: vitalogy
      nodes:
        -
          name: master
          kubeConfig: "#/configs/kubeConfigs/defaultKubeConfig"
          containerConfig: "#/configs/containerConfigs/defaultContainerConfig"
          osConfig:  "#/configs/osConfigs/defaultCoreOsConfig"
configs:
  kubeConfigs:
    -
      name: defaultKubeConfig
      version: v1.5.2
      hyperkubeLocation: gcr.io/google_containers/hyperkube
```

 - An example (partial and simplified) JSON schema:

```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "id": "http://judkins.house/apis/k2/v1",

    "title": "k2 configuration schema",
    "description": "Describes a k2 configuration file.",

    "properties": {
        "version": { "type": "string", "default": "v1" },
        "deployment": { "$ref": "#/definitions/deployment" },
        "configs": { "$ref": "#/definitions/configs" }
    },

    "required": [
        "version"
    ],

    "definitions": {
...
        "configs": {
            "properties": {
                "kubeConfigs": {
                    "title": "Kubernetes configurations",
                    "description": "A list of Kubernetes configurations referenced within the configuration file.",
                    "items": {
                        "$ref": "#/definitions/kubeConfig",
                        "minItems": 1
                    },
                    "type": "array"
                }
            },
            "required": [
                "kubeConfigs"
            ],
            "type": "object"
        },
...
        "kubeConfig": {
            "title": "A Kubernetes configuration",
            "description": "The location and version of a container containing the Kubernetes hyperkube binary.",
            "properties": {
                "name": {
                    "default": "defaultKubeConfig",
                    "description": "Name of the Kubernetes configuration.",
                    "type": "string"
                },
                "hyperkubeLocation": {
                    "default": "gcr.io/google_containers/hyperkube",
                    "description": "Location of the Kubernetes container.",
                    "format": "uri",
                    "type": "string"
                },
                "version": {
                    "default": "v1.5.2",
                    "description": "Version of the hyperkube binary.",
                    "format": "symver",
                    "type": "string"
                }
            },
...
```

## Required

JSON schema already provides a mechanism for marking a property "required". Changes or additions to the properties which are required must result in a new version of the schema. By default additional properties can added to a configuration and it will still be valid against the schema. This can be useful during development, but can also result in Ansible tasks which are dependent on properties which are not documented by the schema. To avoid this we can either write custom code which identifies extra properties, or force the developer to modify the schema when adding properties by setting ```"additionalProperties": false```. The preference is for the later since it results in less code to maintain and we expect properties to added rarely.

## Depreciated

Kuberetes [publishes](https://kubernetes.io/docs/deprecation-policy/) different different depreciation policies for different components. For example, "CLI elements of user-facing components (e.g. kubectl) must function after their announced deprecation for no less than:"

- GA: 1 year or 2 releases (whichever is longer)
- Beta: 3 months or 1 release (whichever is longer)
- Alpha: 0 releases

On the other hand, "CLI elements of admin-facing components (e.g. kubelet) must function after their announced deprecation for no less than":

- GA: 6 months or 1 release (whichever is longer)
- Beta: 3 months or 1 release (whichever is longer)
- Alpha: 0 releases

To depreciate a feature we will extend JSON schema with a depreciated stanza. Custom code will have to be written to print a warning when a feature is depreciated.

NOTES

0) Consider whether nodes and nodepools can/should be merged.
1) Consider splitting up mounts and mountPoints in an analogous way to how k8s splits up volumes and volumeMounts

**Comment by davidewatson** [comment]
# Amendments
 - Consider using the Ansible ```json_query``` filter instead of expanding JSON pointers. For example:

```
# Get a hash map with all ports and names of a cluster:
- name: "Display all server ports and names from cluster1"
  debug: var=item
  with_items: "{{domain_definition|json_query(server_name_cluster1_query)}}"
  vars:
    server_name_cluster1_query: "domain.server[?cluster=='cluster2'].{name: name, port: port}"
```

 - ```version``` properties should not have a ```format``` (e.g. symver). While we might hope that k8s versioning remains consistent, the recent docker change from using MAJOR.MINOR.PATCH versions to using YY.MM versions shows that changes can and do occur.

**Comment by coffeepac** [comment]
@davidewatson I like the proposal as a whole where there is a block of things we care about (k8s deployments) and then blocks of things the k8s deployment cares about.  +1!  I also really like having an easily understood schema for the config file.

I would prefer to expand the JSON pointers instead of using the filter.  I'd like to make consumption of the config straight forward at the expense of config parsing a little more up front work.

the version property should have a format.  I think that while changes like Docker's recent change do occur they are rare and come with an announcement (or should).  Ensuring that we are using a real version for things seems sane.

**Comment by davidewatson** [comment]
You're right about consumption being the important task to make cheap (since it is the most common). My explanation above was sloppy though. The amendment was really about whether we liked the JSON Ref/Pointer syntax. The consensus is that it is ugly, but at least it's a standard: ```      nodePools:
        -
          name: etcd
          count: 5
          kubeConfig: { $ref: "#/definitions/kubeConfigs/default" }
          etcdConfig: { $ref: "#/definitions/etcdConfigs/default" }``` The alternative would be to continue replacing names with stanzas, al la ```      nodePools:
        -
          name: etcd
          count: 5
          kubeConfig: defaultKubeConfig
          etcdConfig: defaultEtcdConfig``` So far no one has defended this approach. Even with documentation it ends up being difficult to reason about without magic.

Fair enough about the version property format. If we only promise backwards compatibility, we can always extend the schema to accept two version formats.

**Comment by davidewatson** [comment]
See #258 for the final config file format.

tl;dr

A default configuration file will look like this:

```
deployment:
  clusters:
    - name: k2recon
      network: 10.32.0.0/12
      dns: 10.32.0.2
      domain: cluster.local
      providerConfig: *defaultAws
      nodePools:
        - name: etcd
          count: 5
          etcdConfig: *defaultEtcd
          containerConfig: *defaultDocker
          osConfig: *defaultCoreOs
          nodeConfig: *defaultAwsEtcdNode
          keyPair: *defaultKeyPair
        - name: etcdEvents
          count: 5
          etcdConfig: *defaultEtcdEvents
          containerConfig: *defaultDocker
          osConfig: *defaultCoreOs
          nodeConfig: *defaultAwsEtcdEventsNode
          keyPair: *defaultKeyPair
        - name: master
          count: 3
          loadBalancerConfig: *defaultLoadBalancer
          apiServerConfig: *defaultApiServer
          kubeConfig: *defaultKube
          containerConfig: *defaultDocker
          osConfig: *defaultCoreOs
          nodeConfig: *defaultAwsMasterNode
          keyPair: *defaultKeyPair
        - name: clusterNodes
          count: 3
          kubeConfig: *defaultKube
          containerConfig: *defaultDocker
          osConfig: *defaultCoreOs
          nodeConfig: *defaultAwsClusterNode
          keyPair: *defaultKeyPair
        - name: specialNodes
          count: 2
          kubeConfig: *defaultKube
          containerConfig: *defaultDocker
          osConfig: *defaultCoreOs
          nodeConfig: *defaultAwsSpecialNode
          keyPair: *defaultKeyPair
      fabricConfig: *defaultCanalFabric
      helmConfig: *defaultHelm
  readiness:
    type: exact
    value: 0
    wait: 600
```

Where the YAML anchors defining things like ```*defaultCoreOs``` are defined before, either within the same file, or possibly within separate files.

**Comment by coffeepac** [comment]
coffeepac to take a look and see what work remains and what kind of priority it has.

**Comment by coffeepac** [comment]
json schema validator is complete.  This ticket was about creating that.  the support and other bits for that ticket are also done.  away, away, away.
